src/vellum/workflows/nodes/core/map_node/tests/test_node.py

```diff
@@ -221,3 +222,56 @@ class TestMapNode(MapNode):
     # AND each item should have run on a different thread
     thread_ids_list = list(thread_ids.values())
     assert len(set(thread_ids_list)) == 3
+
+
+def test_map_node__shared_state_race_condition():
+    processed_items: list[str] = []
+
+    # GIVEN a templating node that processes items
+    class TemplatingNode(BaseNode):
+        item = MapNode.SubworkflowInputs.item
+
+        class Outputs(BaseOutputs):
+            processed_item: str
+
+        def run(self) -> Outputs:
+            processed_item = f"{self.item}!"
+            return self.Outputs(processed_item=processed_item)
+
+    # AND a final output node
+    class FinalOutput(FinalOutputNode[BaseState, str]):
+        class Outputs(FinalOutputNode.Outputs):
+            value = TemplatingNode.Outputs.processed_item
+
+        def run(self) -> Outputs:
+            outputs = super().run()
+            processed_items.append(outputs.value)
```

Comment:
this is what helped me find the bug. This, plus running the node a bunch of times increases the odds that we run into the concurrency issue